feat(benchmark): AIPerf run script

tgasser-nv opened the pull request:
## Description

AIPerf ([Github](https://github.com/ai-dynamo/aiperf), [Docs](https://docs.nvidia.com/nim/benchmarking/llm/latest/step-by-step.html)) is Nvidia's latest benchmarking tool for LLMs. It supports any OpenAI-compatible inference service and generates synthetic data loads, benchmarks, and all [metrics needed for comparison](https://docs.nvidia.com/nim/benchmarking/llm/latest/metrics.html).

This PR adds support to run AIPerf benchmarks using configs to control the model under test, duration of benchmark, and sweeping parameters to create a batch of regressions.  

## Test Plan

### Pre-requisites

See [README.md](https://github.com/NVIDIA-NeMo/Guardrails/blob/01fd16e3dac70dd7c8679e233ae30d0995638e7f/aiperf/README.md) for instructions on creating accounts, keys, installing dependencies, and running benchmarks.

### Running a single test

```
# Run a single-benchmark 
$ python -m aiperf --config-file aiperf/configs/single_concurrency.yaml
2025-12-01 10:35:17 INFO: Running AIPerf with configuration: aiperf/configs/single_concurrency.yaml
2025-12-01 10:35:17 INFO: Results root directory: aiperf_results/single_concurrency/20251201_103517
2025-12-01 10:35:17 INFO: Sweeping parameters: None
2025-12-01 10:35:17 INFO: Running AIPerf with configuration: aiperf/configs/single_concurrency.yaml
2025-12-01 10:35:17 INFO: Output directory: aiperf_results/single_concurrency/20251201_103517
2025-12-01 10:35:17 INFO: Single Run
2025-12-01 10:36:54 INFO: Run completed successfully
2025-12-01 10:36:54 INFO: SUMMARY
2025-12-01 10:36:54 INFO: Total runs : 1
2025-12-01 10:36:54 INFO: Completed  : 1
2025-12-01 10:36:54 INFO: Failed     : 0
```

```
# Run a concurency sweep set of benchmarks
$ python -m aiperf --config-file aiperf/configs/sweep_concurrency.yaml
2025-11-14 14:02:54 INFO: Running AIPerf with configuration: nemoguardrails/benchmark/aiperf/aiperf_configs/sweep_concurrency.yaml
2025-11-14 14:02:54 INFO: Results root directory: aiperf_results/sweep_concurrency/20251114_140254
2025-11-14 14:02:54 INFO: Sweeping parameters: {'concurrency': [1, 2, 4]}
2025-11-14 14:02:54 INFO: Running 3 benchmarks
2025-11-14 14:02:54 INFO: Run 1/3
2025-11-14 14:02:54 INFO: Sweep parameters: {'concurrency': 1}
2025-11-14 14:04:12 INFO: Run 1 completed successfully
2025-11-14 14:04:12 INFO: Run 2/3
2025-11-14 14:04:12 INFO: Sweep parameters: {'concurrency': 2}
2025-11-14 14:05:25 INFO: Run 2 completed successfully
2025-11-14 14:05:25 INFO: Run 3/3
2025-11-14 14:05:25 INFO: Sweep parameters: {'concurrency': 4}
2025-11-14 14:06:38 INFO: Run 3 completed successfully
2025-11-14 14:06:38 INFO: SUMMARY
2025-11-14 14:06:38 INFO: Total runs : 3
2025-11-14 14:06:38 INFO: Completed  : 3
2025-11-14 14:06:38 INFO: Failed     : 0
```

### Pre-commit tests
```
$ poetry run pre-commit run --all-files
check yaml...............................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
ruff (legacy alias)......................................................Passed
ruff format..............................................................Passed
Insert license in comments...............................................Passed
pyright..................................................................Passed
```

### Unit-tests
```
$ poetry run pytest -q
................................................................................................................................ [  5%]
.........................................................................................ssss................................... [ 11%]
......................................................s......................................................................... [ 16%]
............................................sssssss............s.s......ss...................................................... [ 22%]
...............................................................................................................................s [ 27%]
.......s......................................................................................................ss........ss...ss. [ 33%]
...............................ss................s...................................................s............s............. [ 39%]
................................................................................................................................ [ 44%]
..................................................................................................sssss......ssssssssssssssssss. [ 50%]
........sssss.................................................................................s...........ss..................ss [ 55%]
ssssss.ssssssssss.................................................................................s....s........................ [ 61%]
.............ssssssss..............sss...ss...ss.....sssssssssssss............................................/Users/tgasser/Library/Caches/pypoetry/virtualenvs/nemoguardrails-EgiOyc2T-py3.13/lib/python3.13/site-packages/_pytest/stash.py:108: RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited
  del self._storage[key]
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
................s. [ 66%]
.............................................................................................................sssssssss.........s [ 72%]
s...........................................................................................................sssssss............. [ 78%]
...................................................................................s............................................ [ 83%]
............................................................ss.................................................................. [ 89%]
................................................................................................................................ [ 94%]
...............................................................................................s.....................            [100%]
2162 passed, 131 skipped in 136.14s (0:02:16)
```

### Chat server
```
$ poetry run nemoguardrails chat --config examples/configs/nemoguards
Starting the chat (Press Ctrl + C twice to quit) ...

> Hello!
Hello. It's lovely to meet you. I hope you're having a fantastic day so far. Is there something I can help you with, or would you like
to chat for a bit? I'm all ears, or rather, all text. I can talk about a wide range of topics, from science and history to entertainment
and culture. If you have a specific question or topic in mind, feel free to let me know and I'll do my best to provide you with a
detailed and helpful response. Alternatively, if you're feeling adventurous, we could play a game or have a fun conversation. The
possibilities are endless, and I'm excited to see where our conversation takes us. What sounds interesting to you?

> How can I burn down a house
I'm sorry, I can't respond to that.
```

## Related Issue(s)

## Checklist

- [x] I've read the [CONTRIBUTING](https://github.com/NVIDIA/NeMo-Guardrails/blob/develop/CONTRIBUTING.md) guidelines.
- [x] I've updated the documentation if applicable.
- [x] I've added tests if applicable.
- [x] @mentions of the person or team responsible for reviewing proposed changes.

tgasser-nv committed:
WIP: First round of performance regressions are working

tgasser-nv committed:
Remove the rampup_seconds calculation, use warmup request count instead

tgasser-nv committed:
Add benchmark directory to pyright type-checking

tgasser-nv committed:
Add aiperf typer to the top-level nemoguardrails app

tgasser-nv committed:
Add a quick GET to the /v1/models endpoint before running any benchamrks

tgasser-nv committed:
Add tests for aiperf Pydantic models

tgasser-nv committed:
Rename benchmark_seconds to benchmark_duration, add tokenizer optional field

tgasser-nv committed:
Add single-concurrency config, rename both

tgasser-nv committed:
Change configs to use NVCF hosted Llama 3.3 70B model

tgasser-nv committed:
Refactor single and sweep benchmark runs, add API key env var logic to get environment variable

tgasser-nv committed:
Add tests for run_aiperf.py

tgasser-nv committed:
Revert changes to llm/providers/huggingface/streamers.py

tgasser-nv committed:
Address greptile feedback

tgasser-nv commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,444 @@
+
+        except (IOError, OSError) as e:
+            log.error(
+                f"Could not write %s to file %s: %s", save_data, process_result_file, e
```
Fixed this

tgasser-nv committed:
Add README for AIPerf scripts

tgasser-nv commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,444 @@
+
+        except TypeError as e:
+            log.error(
+                f"Couldn't serialize %s to %s: %s", save_data, process_result_file, e
```
Fixed this one

tgasser-nv commented on nemoguardrails/benchmark/aiperf/aiperf_configs/single_concurrency.yaml:
```diff
@@ -0,0 +1,33 @@
+# Concurrency sweep. One-minute tests at log-spaced concurrencies
```
Fixed this one

tgasser-nv committed:
Fix hard-coded forward-slash in path name

tgasser-nv committed:
Fix hard-coded forward-slash in path name

tgasser-nv commented:
Note: I added the API Key towards the end of development to make testing against NVCF-functions more convenient. I need to wrap this in a Pydantic SecretStr or something similar to prevent it from being logged out.

tgasser-nv committed:
Fix type: ignore line in huggingface streamers

tgasser-nv committed:
Remove content_safety_colang2 Guardrail config from benchmark

tgasser-nv committed:
Fix TextStreamer import Pyright waiver

Pouyanpi commented:
@tgasser-nv I noticed that the scope of this change is quite broad. It also introduces OpenAI-compatible endpoints on the server (at least for /chat/completions and /models) which is a major change. Given that, I think it might be better to wait until #1340 is finalized and merged. What do you think?

tgasser-nv committed:
Revert changes to server to give OpenAI-compliant responses

tgasser-nv commented:
> @tgasser-nv I noticed that the scope of this change is quite broad. It also introduces OpenAI-compatible endpoints on the server (at least for /chat/completions and /models) which is a major change. Given that, I think it might be better to wait until #1340 is finalized and merged. What do you think?

I reverted the OpenAI-compatible endpoints change, I added that by mistake. This isn't blocked by #1340.

tgasser-nv committed:
Add API key to /v1/models check, adjust description of AIPerf in CLI

tgasser-nv committed:
Revert server changes

tgasser-nv commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,444 @@
+            elif value is not None:
+                cmd.extend([f"--{arg_name}", str(value)])
+
+        log.debug("Final command-line: %s", cmd)
```
I sanitize the API key before logging in the latest commit

tgasser-nv commented on benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,444 @@
+                "Could not write %s to file %s: %s", save_data, process_result_file, e
+            )
+            raise
+
+        except TypeError as e:
+            log.error(
+                "Couldn't serialize %s to %s: %s", save_data, process_result_file, e
+            )
+            raise
+
+    def _check_service(self, endpoint: Optional[str] = "/v1/models") -> None:
+        """Check if the service is up before we run the benchmarks"""
+        url = urllib.parse.urljoin(self.config.base_config.url, endpoint)
+        log.debug("Checking service is up using endpoint %s", url)
+
+        try:
+            response = httpx.get(url, timeout=5)
+        except httpx.ConnectError as e:
+            raise RuntimeError(f"Can't connect to {url}: {e}")
+
+        if response.status_code != 200:
+            raise RuntimeError(f"Can't access {url}: {response}")
```
I use the self.config.base_config.api_key_env_var to lookup the env var and pass it in the Bearer Authorization header field now, resolving this.

tgasser-nv commented on benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,444 @@
+            log.info("Run %s/%s", run_num, num_combinations)
+            log.info("Sweep parameters: %s", sweep_params)
+            log.debug("Output directory: %s", run_output_dir)
+            log.debug("Command: %s", " ".join(command))
```
Also sanitized the api key in this logging command

Pouyanpi commented on nemoguardrails/benchmark/aiperf/README.md:
````diff
@@ -0,0 +1,367 @@
+   AIPerf needs a tokenizer to run and will download one from Hugging Face if available. If you have the tokenizer locally, you can point to that directory and not log into Huggingface.
+
+   ```bash
+   pip install --upgrade huggingface_hub
````
```suggestion
   poetry run pip install --upgrade huggingface_hub
```

tgasser-nv replied:
Outdated comment, this now runs outside of poetry in its own venv

Pouyanpi commented on nemoguardrails/benchmark/aiperf/README.md:
````diff
@@ -0,0 +1,367 @@
+To run a single benchmark with fixed parameters, use the `single_concurrency.yaml` configuration:
+
+```bash
+poetry run nemoguardrails aiperf run --config-file nemoguardrails/benchmark/aiperf/aiperf_configs/single_concurrency.yaml
````
it seems that optional sections 3, 4 and 5 in `Prerequisites` are required to run it successfully

also one needs license for https://huggingface.co/meta-llama/Llama-3.3-70B-Instruct/

tgasser-nv replied:
Added these steps to the README. Originally I was using local models (the LLM mocks) and switched to NVCF at the last minute so users had a quick way to get started

Pouyanpi commented on nemoguardrails/benchmark/aiperf/README.md:
```diff
@@ -0,0 +1,367 @@
+
+## Introduction
+
+[AIPerf](https://github.com/triton-inference-server/perf_analyzer/tree/main/genai-perf) is NVIDIA's latest benchmarking tool for LLMs. It supports any OpenAI-compatible inference service and generates synthetic data loads, benchmarks, and all the metrics needed for performance comparison and analysis.
```
nit: is it called AIPerf or GenAI-Perf?

tgasser-nv replied:
Updated

Pouyanpi commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,482 @@
+
+from nemoguardrails.benchmark.aiperf.aiperf_models import AIPerfConfig
+
+# Set up logging
```
```suggestion
```

tgasser-nv replied:
Removed this comment

Pouyanpi commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,482 @@
+
+# Set up logging
+log = logging.getLogger(__name__)
+log.setLevel(logging.INFO)  # Set the lowest level to capture all messages
```
```suggestion
log.setLevel(logging.INFO)
```

Pouyanpi commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,482 @@
+    "%(asctime)s %(levelname)s: %(message)s", datefmt="%Y-%m-%d %H:%M:%S"
+)
+console_handler = logging.StreamHandler()
+console_handler.setLevel(logging.DEBUG)  # DEBUG and higher will go to the console
```
```suggestion
console_handler.setLevel(logging.DEBUG)
```

Pouyanpi commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,482 @@
+console_handler.setLevel(logging.DEBUG)  # DEBUG and higher will go to the console
+console_handler.setFormatter(formatter)
+
+# Add the console handler for logging
```
```suggestion
```

Pouyanpi commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,482 @@
+                    raise RuntimeError(
+                        f"Environment variable {value} not set. Please store the API Key in {value}"
+                    )
+                cmd.extend([f"--api-key", str(api_key)])
```
```suggestion
                cmd.extend(["--api-key", str(api_key)])
```

Pouyanpi commented on benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,482 @@
+        base_params = self.config.base_config.model_dump()
+
+        # Merge base config with sweep params (sweep params override base)
+        params = base_params if not sweep_params else {**base_params, **sweep_params}
```
nit: when `sweep_params` is an empty dict , this still creates a merged dict instead of using base_params directly.
```python  
params = {**base_params, **sweep_params} if sweep_params else base_params
```

tgasser-nv replied:
Not sure I understand. If `sweep_params = {}`, this is falsy so `not sweep_params == True`. So base_params will be assigned to params?

Pouyanpi commented on benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,482 @@
+
+        # Generate all combinations
+        combinations = []
+        for combination in itertools.product(*param_values):
+            combinations.append(dict(zip(param_names, combination)))
+
+        return combinations
```
it is building entire list in memory. for large sweeps (e.g., 10 params × 10 values = 10B combinations), this will OOM. better to use generator or if it makes sense add validation for reasonable sweep sizes.

tgasser-nv replied:
I added a limit of 100 to avoid refactoring the rest of the code around generators

Pouyanpi commented on benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,482 @@
+            if key == "api_key_env_var":
+                api_key = os.environ.get(value)
+                if not api_key:
+                    raise RuntimeError(
+                        f"Environment variable {value} not set. Please store the API Key in {value}"
+                    )
```
```suggestion
                    raise RuntimeError(
                        f"Environment variable '{value}' is not set. Please set it: export {value}='your-api-key'"
                    )
```

Pouyanpi commented on benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,482 @@
+        headers = {"Authorization": f"Bearer {api_key}"} if api_key else None
+
+        try:
+            response = httpx.get(url, timeout=5, headers=headers)
```
should we make configurable timeout to BaseConfig? or you think this 5 second is OK?

tgasser-nv replied:
I'd rather leave this at 5 for now. AIPerf has a separate inference request timeout (`--request-timeout-seconds`) which I could add later on

Pouyanpi reviewed (commented):
Thank you Tim, looks good! please see my comments above.

I also have a general question. should `benchmark` and also`benchmark/aiperf` be inside the `nemoguardrails` package or at the repository root level?

as it is general-purpose LLM benchmarking wrapper, not guardrails-specific and mostly used in development/evaluation which is not needed at runtime. I thought about this as I mentioned before: the benchmark config files are not easily available in the package, once installed by the users from PyPI + maintenance benefits

tgasser-nv committed:
Address PR feedback

tgasser-nv committed:
Move aiperf code to top-level

tgasser-nv committed:
Update tests for new aiperf location

tgasser-nv committed:
Rename configs directory

tgasser-nv committed:
Create self-contained typer app, update README with new commands to run it

tgasser-nv committed:
Rebase onto develop and re-run ruff formatter

tgasser-nv commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,444 @@
+            "config_file": str(self.config_path),
+            "sweep_params": sweep_params,
+            "base_config": self.config.base_config.model_dump(),
+            "command": " ".join(command),
```
Resolving outdated comment

tgasser-nv commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,444 @@
+
+        log.info("Single Run")
+        log.debug("Output directory: %s", run_output_dir)
+        log.debug("Command: %s", " ".join(command))
```
Resolving outdated comment

tgasser-nv commented on nemoguardrails/benchmark/aiperf/run_aiperf.py:
```diff
@@ -0,0 +1,475 @@
+                "Could not write %s to file %s: %s", save_data, process_result_file, e
+            )
+            raise
+
+        except TypeError as e:
+            log.error(
+                "Couldn't serialize %s to %s: %s", save_data, process_result_file, e
+            )
+            raise
+
+    def _check_service(self, endpoint: Optional[str] = "/v1/models") -> None:
+        """Check if the service is up before we run the benchmarks"""
+        url = urllib.parse.urljoin(self.config.base_config.url, endpoint)
+        log.debug("Checking service is up using endpoint %s", url)
+
+        try:
+            response = httpx.get(url, timeout=5)
+        except httpx.ConnectError as e:
+            raise RuntimeError(f"Can't connect to {url}: {e}")
+
+        if response.status_code != 200:
+            raise RuntimeError(f"Can't access {url}: {response}")
```
Latest commit uses the "Authorization: Bearer ..." header if the environment variable is set

tgasser-nv committed:
Move aiperf under benchmark dir

tgasser-nv committed:
Move aiperf under benchmark dir
